Clean edit pass on attributes

BillWagner opened the pull request:
Fixes #42850:  Add a new page to reference attributes that cause the compiler to emit different IL modifiers instead of putting the attribute in the emitted IL.
Fixes #43109: Most of these have been added to the general page. A few are expressed in source differently, and were added to the new pseudo-attributes page.
Fixes #43110: Add a list of attributes that are written by the compiler that shouldn't be added in users' source code.


<!-- PREVIEW-TABLE-START -->

---

#### Internal previews

| 📄 File | 🔗 Preview link |
|:--|:--|
| [docs/csharp/language-reference/attributes/general.md](https://github.com/dotnet/docs/blob/b6bb555ff3c685b1ffeb0aa6c49a6dd1d8edbf15/docs/csharp/language-reference/attributes/general.md) | [Miscellaneous attributes interpreted by the C# compiler](https://review.learn.microsoft.com/en-us/dotnet/csharp/language-reference/attributes/general?branch=pr-en-us-44513) |
| [docs/csharp/language-reference/attributes/pseudo-attributes.md](https://github.com/dotnet/docs/blob/b6bb555ff3c685b1ffeb0aa6c49a6dd1d8edbf15/docs/csharp/language-reference/attributes/pseudo-attributes.md) | ["Attributes interpreted by the compiler: Pseudo-attributes"](https://review.learn.microsoft.com/en-us/dotnet/csharp/language-reference/attributes/pseudo-attributes?branch=pr-en-us-44513) |
| [docs/csharp/language-reference/toc.yml](https://github.com/dotnet/docs/blob/b6bb555ff3c685b1ffeb0aa6c49a6dd1d8edbf15/docs/csharp/language-reference/toc.yml) | [docs/csharp/language-reference/toc](https://review.learn.microsoft.com/en-us/dotnet/csharp/language-reference/toc?branch=pr-en-us-44513) |
| [docs/csharp/misc/cs0612.md](https://github.com/dotnet/docs/blob/b6bb555ff3c685b1ffeb0aa6c49a6dd1d8edbf15/docs/csharp/misc/cs0612.md) | [Compiler warning (level 1) CS0612](https://review.learn.microsoft.com/en-us/dotnet/csharp/misc/cs0612?branch=pr-en-us-44513) |
| [docs/csharp/misc/cs0619.md](https://github.com/dotnet/docs/blob/b6bb555ff3c685b1ffeb0aa6c49a6dd1d8edbf15/docs/csharp/misc/cs0619.md) | ["Compiler Error CS0619"](https://review.learn.microsoft.com/en-us/dotnet/csharp/misc/cs0619?branch=pr-en-us-44513) |
| [docs/csharp/specification/toc.yml](https://github.com/dotnet/docs/blob/b6bb555ff3c685b1ffeb0aa6c49a6dd1d8edbf15/docs/csharp/specification/toc.yml) | [docs/csharp/specification/toc](https://review.learn.microsoft.com/en-us/dotnet/csharp/specification/toc?branch=pr-en-us-44513) |
| [docs/csharp/versioning.md](https://github.com/dotnet/docs/blob/b6bb555ff3c685b1ffeb0aa6c49a6dd1d8edbf15/docs/csharp/versioning.md) | [Versioning in C\#](https://review.learn.microsoft.com/en-us/dotnet/csharp/versioning?branch=pr-en-us-44513) |
| [docs/csharp/whats-new/csharp-12.md](https://github.com/dotnet/docs/blob/b6bb555ff3c685b1ffeb0aa6c49a6dd1d8edbf15/docs/csharp/whats-new/csharp-12.md) | [What's new in C# 12](https://review.learn.microsoft.com/en-us/dotnet/csharp/whats-new/csharp-12?branch=pr-en-us-44513) |
| [docs/fundamentals/networking/http/httpclient-guidelines.md](https://github.com/dotnet/docs/blob/b6bb555ff3c685b1ffeb0aa6c49a6dd1d8edbf15/docs/fundamentals/networking/http/httpclient-guidelines.md) | [Guidelines for using HttpClient](https://review.learn.microsoft.com/en-us/dotnet/fundamentals/networking/http/httpclient-guidelines?branch=pr-en-us-44513) |
| [docs/fundamentals/runtime-libraries/preview-apis.md](https://github.com/dotnet/docs/blob/b6bb555ff3c685b1ffeb0aa6c49a6dd1d8edbf15/docs/fundamentals/runtime-libraries/preview-apis.md) | [docs/fundamentals/runtime-libraries/preview-apis](https://review.learn.microsoft.com/en-us/dotnet/fundamentals/runtime-libraries/preview-apis?branch=pr-en-us-44513) |


<!-- PREVIEW-TABLE-END -->

BillWagner committed:
Add page for custom attributes

Fixes #42850:  Add a new page to reference attributes that cause the compiler to emit different IL modifiers instead of putting the attribute in the emitted IL.

BillWagner committed:
Add additional attributes

Fixes #43109

Most of these have been added to the general page. A few are expressed in source differently, and were added to the new pseudo-attributes page.

BillWagner committed:
Fixes #43110

Add a list of attributes that are written by the compiler that shouldn't be added in users' source code.

BillWagner committed:
Add any atributes to displayName

Some attributes weren't in the `displayName` element for the TOC. Review all of them and add them again.

BillWagner committed:
edit pass

BillWagner committed:
Fix build warnings

Mostly fix anchors.

The one C# 9 spec was never intended to be published.

gewarren commented on docs/csharp/language-reference/attributes/general.md:
```diff
@@ -58,15 +65,19 @@ The string provided as the first argument to the attribute constructor is displa
 
 > [!WARNING]
-> Experimental features are subject to changes. The APIs may change, or they may be removed in future updates. Including experimental features is a way for library authors to get feedback on ideas and concepts for future development. Use extreme caution when using any feature marked as experimental.
+> Experimental features are subject to changes. The APIs can change, or they can be removed in future updates. Including experimental features is a way for library authors to get feedback on ideas and concepts for future development. Use extreme caution when using any feature marked as experimental.
```
Once https://github.com/dotnet/docs/pull/44436 is merged, we should interlink between https://learn.microsoft.com/en-us/dotnet/fundamentals/runtime-libraries/preview-apis#experimentalattribute and here.

BillWagner replied:
I responded to all the other feedback. Should I merge first, or do you want to?

gewarren replied:
I can't merge until I get an approving review.

BillWagner replied:
I reviewed and approved it.

gewarren commented on docs/csharp/language-reference/attributes/general.md:
```diff
@@ -242,6 +253,22 @@ Overload resolution considers the two methods equally good for some argument typ
 
+## EnumeratorCancellation attribute
+
+The <xref:System.Runtime.CompilerServices.EnumeratorCancellationAttribute?displayProperty=nameWithType> attribute specifies which parameter should receive the cancellation token from <xref:System.Collections.Generic.IAsyncEnumerable`1.GetAsyncEnumerator(System.Threading.CancellationToken)?displayProperty=fullName> API. It's part of the infrastructure for the [async streams](../../asynchronous-pro ...
```
```suggestion
The <xref:System.Runtime.CompilerServices.EnumeratorCancellationAttribute?displayProperty=nameWithType> attribute specifies which parameter should receive the cancellation token from the <xref:System.Collections.Generic.IAsyncEnumerable`1.GetAsyncEnumerator(System.Threading.CancellationToken)?displayProperty=fullName> API. It's part of the infrastructure for the [async streams](../../asynchronous-programming/generate-consume-asynchronous-stream.md) feature.
```

gewarren commented on docs/csharp/language-reference/attributes/general.md:
```diff
@@ -242,6 +253,22 @@ Overload resolution considers the two methods equally good for some argument typ
+
+## CollectionBuilder attribute
+
+The <xref:System.Runtime.CompilerServices.CollectionBuilderAttribute?displayProperty=nameWithType> specifies a method that builds an instance of a collection type from a [collection expression](../operators/collection-expressions.md). You use this attribute to specify a method that builds the collection. The compiler generates code to call that method when a collection expression is converted to  ...
```
```suggestion
The <xref:System.Runtime.CompilerServices.CollectionBuilderAttribute?displayProperty=nameWithType> attribute specifies a method that builds an instance of a collection type from a [collection expression](../operators/collection-expressions.md). You use this attribute to specify a method that builds the collection. The compiler generates code to call that method when a collection expression is converted to that type.
```

gewarren commented on docs/csharp/language-reference/attributes/pseudo-attributes.md:
```diff
@@ -0,0 +1,49 @@
+---
+# Custom attributes that generate flags or options in the Intermediate Language (IL) output
+
+You add these attributes to your code for the compiler to emit specified Intermediate Language (IL) modifier. These attributes instruct the compiler to include the corresponding IL modifier in the output. These IL modifiers can be controlled in C# using these attributes:
```
```suggestion
You add these attributes to your code for the compiler to emit a specified Intermediate Language (IL) modifier. These attributes instruct the compiler to include the corresponding IL modifier in the output.
```

gewarren replied:
Seemed a bit repetitive so I removed the last sentence.

gewarren commented on docs/csharp/language-reference/attributes/pseudo-attributes.md:
```diff
@@ -0,0 +1,49 @@
+
+You add these attributes to your code for the compiler to emit specified Intermediate Language (IL) modifier. These attributes instruct the compiler to include the corresponding IL modifier in the output. These IL modifiers can be controlled in C# using these attributes:
+
+- <xref:System.Runtime.InteropServices.ComImportAttribute?displayProperty=fullName>: Specifies the `import` modifier.
```
This might be nice as a table with the attribute in one column, the modifier in the second, and perhaps comments in a third column. Just a suggestion though.

gewarren commented on docs/csharp/language-reference/attributes/pseudo-attributes.md:
```diff
@@ -0,0 +1,49 @@
+- <xref:System.Runtime.InteropServices.DllImportAttribute?displayProperty=fullName>: Specifies the `pinvokeimpl` modifier. You can add options listed in the constructor.
+- <xref:System.Runtime.InteropServices.FieldOffsetAttribute?displayProperty=fullName>: Specifies the `.field` offset modifier.
+- <xref:System.Runtime.InteropServices.MarshalAsAttribute>: Specifies the IL `marshal` modifier. You can set options listed in the constructor.
+4- <xref:System.Runtime.CompilerServices.MethodImplAttribute?displayProperty=fullName>: maps to the IL `flag` modifier. Constructor arguments specify specific named flags such as `aggressiveinlining` or `forwardref`. These flags also specify the `native`, `managed`, or `optil` modifiers for the <xref:System.Runtime.CompilerServices.MethodCodeType?displayProperty=fullName> field.
```
```suggestion
4- <xref:System.Runtime.CompilerServices.MethodImplAttribute?displayProperty=fullName>: Maps to the IL `flag` modifier. Constructor arguments specify specific named flags such as `aggressiveinlining` or `forwardref`. These flags also specify the `native`, `managed`, or `optil` modifiers for the <xref:System.Runtime.CompilerServices.MethodCodeType?displayProperty=fullName> field.
```

gewarren commented on docs/csharp/language-reference/attributes/pseudo-attributes.md:
```diff
@@ -0,0 +1,49 @@
+
+Some of these custom attributes are applied using other C# syntax rather than adding the attribute to your source code.
+
+- <xref:System.Runtime.InteropServices.DefaultParameterValueAttribute?displayProperty=fullName>: Specifies the default value for the parameter. Use the [default parameter syntax](../../methods.md#optional-parameters-and-arguments)
```
```suggestion
- <xref:System.Runtime.InteropServices.DefaultParameterValueAttribute?displayProperty=fullName>: Specifies the default value for the parameter. Use the [default parameter syntax](../../methods.md#optional-parameters-and-arguments).
```

gewarren commented on docs/csharp/language-reference/attributes/pseudo-attributes.md:
```diff
@@ -0,0 +1,49 @@
+
+- <xref:System.Runtime.InteropServices.DefaultParameterValueAttribute?displayProperty=fullName>: Specifies the default value for the parameter. Use the [default parameter syntax](../../methods.md#optional-parameters-and-arguments)
+- <xref:System.Runtime.InteropServices.InAttribute?displayProperty=fullName>: Specifies the IL `[in]` modifier. Use the [`in`](../keywords/method-parameters.md#in-parameter-modifier) or [`ref readonly`](../keywords/method-parameters.md#ref-readonly-modifier).
+- <xref:System.Runtime.CompilerServices.SpecialNameAttribute?displayProperty=fullName>: Specifies the IL `specialname` modifier. The compiler automatically uses the for methods that require the `specialname` modifier.
```
"the for methods" is correct?

gewarren commented on docs/csharp/language-reference/attributes/pseudo-attributes.md:
```diff
@@ -0,0 +1,49 @@
+- <xref:System.Runtime.InteropServices.DefaultParameterValueAttribute?displayProperty=fullName>: Specifies the default value for the parameter. Use the [default parameter syntax](../../methods.md#optional-parameters-and-arguments)
+- <xref:System.Runtime.InteropServices.InAttribute?displayProperty=fullName>: Specifies the IL `[in]` modifier. Use the [`in`](../keywords/method-parameters.md#in-parameter-modifier) or [`ref readonly`](../keywords/method-parameters.md#ref-readonly-modifier).
+- <xref:System.Runtime.CompilerServices.SpecialNameAttribute?displayProperty=fullName>: Specifies the IL `specialname` modifier. The compiler automatically uses the for methods that require the `specialname` modifier.
+- <xref:System.Runtime.InteropServices.UnmanagedCallersOnlyAttribute?displayProperty=nameWithType>: This attribute is required for the `delegate*` feature. The compiler adds it to any [`delegate*`](../unsafe-code.md#function-pointers) requires its use.
```
```suggestion
- <xref:System.Runtime.InteropServices.UnmanagedCallersOnlyAttribute?displayProperty=nameWithType>: This attribute is required for the `delegate*` feature. The compiler adds it to any [`delegate*`](../unsafe-code.md#function-pointers) that requires its use.
```

gewarren commented on docs/csharp/language-reference/attributes/pseudo-attributes.md:
```diff
@@ -0,0 +1,49 @@
+- <xref:System.Runtime.CompilerServices.IsUnmanagedAttribute?displayProperty=fullName> - This attribute specifies the `unmanaged` constraint on a type parameter.
+- <xref:System.Runtime.CompilerServices.NullableAttribute?displayProperty=fullName>, <xref:System.Runtime.CompilerServices.NullableContextAttribute?displayProperty=fullName>, <xref:System.Runtime.CompilerServices.NullablePublicOnlyAttribute?displayProperty=fullName>: These attributes encode nullable annotations in your source code.
+- <xref:System.ParamArrayAttribute?displayProperty=fullName>: This attribute encodes the `params` modifier on array parameters.
+- <xref:System.Runtime.CompilerServices.ParamCollectionAttribute?displayProperty=fullName> This attribute encodes the `params` on non-array parameters.
```
```suggestion
- <xref:System.Runtime.CompilerServices.ParamCollectionAttribute?displayProperty=fullName> This attribute encodes the `params` modifier on non-array parameters.
```

gewarren commented on docs/csharp/language-reference/attributes/pseudo-attributes.md:
```diff
@@ -0,0 +1,49 @@
+- <xref:System.Runtime.CompilerServices.RequiredMemberAttribute?displayProperty=fullName>: This attribute indicates that the `required` modifier was placed on a member declaration. It's the encoding of the [required members](../keywords/required.md) language feature.
+- <xref:System.Runtime.CompilerServices.TupleElementNamesAttribute?displayProperty=fullName>: This attribute encodes tuple element names used in signatures.
+
+In addition, the compiler can generate a declaration for other attributes used internally. For this reason, you should assume other attributes in the <xref:System.Runtime.CompilerServices> namespace. Some aren't in the .NET Runtime. Instead, the compiler synthesizes a definition for an `internal` type declaration in any assembly where the attribute is needed.
```
Second sentence is incomplete.

BillWagner committed:
Apply suggestions from code review

Co-authored-by: Genevieve Warren <24882762+gewarren@users.noreply.github.com>

BillWagner committed:
Respond to feedback.

BillWagner committed:
rebase and merbe

BillWagner committed:
link between attributes

Link between the C# attribute page and the .NET fundamentals article on experimental attributes in the runtime.

BillWagner committed:
build warning

jnm2 commented on docs/csharp/language-reference/attributes/pseudo-attributes.md:
```diff
@@ -0,0 +1,55 @@
+| <xref:System.Runtime.CompilerServices.SpecialNameAttribute?displayProperty=fullName>             | Specifies the IL `specialname` modifier. The compiler automatically adds this modifier for methods that require it. |
+| <xref:System.Runtime.InteropServices.UnmanagedCallersOnlyAttribute?displayProperty=nameWithType> | This attribute is required for the `delegate*` feature. The compiler adds it to any [`delegate*`](../unsafe-code.md#function-pointers) that requires its use. |
+
+The following custom attributes are generally disallowed in C# source. They're listed here to aid library authors who use reflection, and to ensure you don't create custom attributes with the same name.
```
Though, creating custom attributes with the same name in a different namespace is okay, and custom attributes don't go in the System namespace.

jnm2 commented on docs/csharp/language-reference/attributes/pseudo-attributes.md:
```diff
@@ -0,0 +1,55 @@
+| <xref:System.Runtime.InteropServices.DefaultParameterValueAttribute?displayProperty=fullName>    | Specifies the default value for the parameter. Use the [default parameter syntax](../../methods.md#optional-parameters-and-arguments). |
+| <xref:System.Runtime.InteropServices.InAttribute?displayProperty=fullName>                       | Specifies the IL `[in]` modifier. Use the [`in`](../keywords/method-parameters.md#in-parameter-modifier) or [`ref readonly`](../keywords/method-parameters.md#ref-readonly-modifier). |
+| <xref:System.Runtime.CompilerServices.SpecialNameAttribute?displayProperty=fullName>             | Specifies the IL `specialname` modifier. The compiler automatically adds this modifier for methods that require it. |
+| <xref:System.Runtime.InteropServices.UnmanagedCallersOnlyAttribute?displayProperty=nameWithType> | This attribute is required for the `delegate*` feature. The compiler adds it to any [`delegate*`](../unsafe-code.md#function-pointers) that requires its use. |
```
`[UnmanagedCallersOnly]` is actually something the user must apply in code in order to use the `unmanaged` calling convention specifically with the `delegate*` feature:

```cs
using System.Runtime.InteropServices;

unsafe
{
    delegate* unmanaged<void> someFuncPointer = &SomeMethod;
}

[UnmanagedCallersOnly] // Compilation fails if this is removed
static void SomeMethod() { }
```

jnm2 commented on docs/csharp/language-reference/attributes/pseudo-attributes.md:
```diff
@@ -0,0 +1,55 @@
+| <xref:System.Runtime.CompilerServices.MethodImplAttribute?displayProperty=fullName>  | `flag`          | Constructor arguments specify specific named flags such as `aggressiveinlining` or `forwardref`. These flags also specify the `native`, `managed`, or `optil` modifiers for the <xref:System.Runtime.CompilerServices.MethodCodeType?displayProperty=fullName> field. |
+| <xref:System.NonSerializedAttribute?displayProperty=fullName>                        | `notserialized` |           |
+| <xref:System.Runtime.InteropServices.OptionalAttribute?displayProperty=fullName>     | `[opt]`         |           |
+| <xref:System.Runtime.InteropServices.OutAttribute?displayProperty=fullName>          | `[out]`         |           |
```
This one could have a note parallel to the note added for `in`, saying "Use the `out` modifier."
